src/lightning/pytorch/callbacks/throughput_monitor.py

```diff
@@ -0,0 +1,176 @@
+    | `time/val`               | Total elapsed validation time                                                         |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `time/total`             | Total elapsed time (time/train + time/val)                                            |
+    +--------------------------+---------------------------------------------------------------------------------------+
+
+    Notes:
+        - The implementation assumes that devices are homogeneous as it normalizes by the world size.
+        - items_per_sec, flops_per_sec and MFU do not account for padding if present. We suggest using
+            samples_per_sec or batches_per_sec to measure throughput under this circumstance.
+
+    Args:
+        length_fn: A function to compute the number of items in a sample given a batch.
+        batch_size_fn: A function to compute the number of samples given a batch.
+        window_size: Number of batches to use for a rolling average of throughput.
+        time_unit: Time unit to use for `time` logging.
+
+    """
+
+    def __init__(self, length_fn: Callable[[Any], int], batch_size_fn: Callable[[Any], int], **kwargs: Any) -> None:
+        super().__init__()
+        self._monitor: Optional[_ThroughputMonitorBase] = None
+        self._kwargs = kwargs
+        self.length_fn = length_fn
+        self.batch_size_fn = batch_size_fn
+        self._eval_t0 = 0.0
+        self._train_t0 = 0.0
+        self._total_lengths = 0
+
+    def setup(self, trainer: "Trainer", pl_module: "LightningModule", stage: str) -> None:
+        if self._monitor is not None:
+            return  # already setup
+        dtype = _plugin_to_compute_dtype(trainer.precision_plugin)
+        flops_available = _get_flops_available(trainer.strategy.root_device, dtype)
+        if flops_available is not None and not hasattr(pl_module, "flops_per_batch"):
+            rank_zero_info(
+                "When using the `ThroughputMonitor`, you need to define a `flops_per_batch` attribute or property"
+                f" in {pl_module} to compute the FLOPs."
+            )
+        self._monitor = _ThroughputMonitorBase(flops_available, trainer.loggers, **self._kwargs)
+
+    @rank_zero_only
+    def on_train_start(self, trainer: "Trainer", pl_module: "LightningModule") -> None:
+        self._train_t0 = time.perf_counter()
+
+    @rank_zero_only
+    @torch.no_grad()
+    def on_train_batch_end(
+        self, trainer: "Trainer", pl_module: "LightningModule", outputs: Any, batch: Any, batch_idx: int
+    ) -> None:
+        train_elapsed = time.perf_counter() - self._train_t0
+        self._total_lengths += self.length_fn(batch)
+        if trainer.fit_loop._should_accumulate():
+            return
+        flops_per_batch = pl_module.flops_per_batch if hasattr(pl_module, "flops_per_batch") else None
```

Comment:
I don't see us setting this attribute anywhere, is this the user?

Reply:
Yes. It's currently configured to show a warning: https://github.com/Lightning-AI/lightning/pull/18848/files#diff-f96ff2655d6e89aae7394d25969a2b1c207532b94ce2ecd4712f6b1e3021fdffR108-R112